### PR TITLE
Add rule for enforcing directive restrictions

### LINKED ISSUE #23: Restrict to Elements and Attributes (Y074)
https://github.com/johnpapa/angularjs-styleguide/edit/master/README.md

- When creating a directive that makes sense as a stand-alone element, allow restrict `E` (custom element) and optionally restrict `A` (custom attribute). Generally, if it could be its own control, `E` is appropriate. General guideline is allow `EA` but lean towards implementing as an element when it's stand-alone and as an attribute when it enhances its existing DOM element.

    *Why?*: It makes sense.

    *Why?*: While we can allow the directive to be used as a class, if the directive is truly acting as an element it makes more sense as an element or at least as an attribute.

    Note: EA is the default for AngularJS 1.3 +

  ```html
  <!-- avoid -->
  <div class="my-calendar-range"></div>
  ```

  ```javascript
  /* avoid */
  angular
      .module('app.widgets')
      .directive('myCalendarRange', myCalendarRange);

  function myCalendarRange() {
      var directive = {
          link: link,
          templateUrl: '/template/is/located/here.html',
          restrict: 'C'
      };
      return directive;

      function link(scope, element, attrs) {
        /* */
      }
  }
  ```

  ```html
  <!-- recommended -->
  <my-calendar-range></my-calendar-range>
  <div my-calendar-range></div>
  ```

  ```javascript
  /* recommended */
  angular
      .module('app.widgets')
      .directive('myCalendarRange', myCalendarRange);

  function myCalendarRange() {
      var directive = {
          link: link,
          templateUrl: '/template/is/located/here.html',
          restrict: 'EA'
      };
      return directive;

      function link(scope, element, attrs) {
        /* */
      }
  }
  ```

### PR DESCRIPTION
**Pull request by remcohaszing**
There are some corner cases. For example when setting a property named `restrict` to another object inside of a directive declaration. I think this is rare though.

Closes #23

### CODE REVIEW

File: rules/directive-restrict.js [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+module.exports = function(context) {
+    var utils = require('./utils/utils');
+
+    var options = context.options[0] || {};
+    var restrictOpt = options.restrict || 'AE';
+    var explicitRestrict = options.explicit === 'always';
+    var restrictChars = [];
+
+    restrictOpt.split('').forEach(function(char) {
+        if ('ACEM'.indexOf(char) === -1) {
```
Comment (tilmanpotthof) [inline rules/directive-restrict.js]: This code is unnecessary, because you already check the options with the json schema. That's also the reason why the code coverage decreases.

var restrictedChars = restrictOps.split('');

**Comment by tilmanpotthof** [comment]
@remcohaszing I personally think those corner cases are negligible.

File: rules/directive-restrict.js (hunk 1 above)
Comment (remcohaszing) [inline rules/directive-restrict.js]: I figured users could potentially misconfigure the acceptible restrict as for example `AEA`. I don't know how to check this using just a pattern. Should I just assume users won't do this?

**Comment by Gillespie59** [comment]
I agree. You can have a look to the issue #130 after this rule. It is quiet similqar to this one ;)

File: rules/directive-restrict.js (hunk 1 above)
Comment (tilmanpotthof) [inline rules/directive-restrict.js]: Ok, I understand your intention, but these lines of code only filter all chars not in 'ACEM' what is already checked by the regex, so line 13 can not be executed ever and that's why the line can not be covered by any test.

BUT I think your ACA corner case should be checked. And it can be checked with a regex.

/^A?C?E?M?$/

https://www.debuggex.com/r/nJTGB4V9fYnABGKt
